crates/usvg/src/lib.rs

```diff
@@ -69,6 +81,125 @@ pub trait TreeWriting {
 
 impl TreeWriting for usvg_tree::Tree {
     fn to_string(&self, opt: &XmlOptions) -> String {
-        writer::convert(self, opt)
+        let used_ids = get_ids(self);
+        let mut id_generator = IdGenerator::new(used_ids);
+
+        let mut writer_context = WriterContext {
+            tree: &self,
+            opt,
+            text_path_map: Vec::new(),
+            id_generator: &mut id_generator,
+        };
+
+        writer::convert(&mut writer_context)
+    }
+}
+
+/// A generator for new IDs.
+#[derive(Default)]
+pub(crate) struct IdGenerator {
+    /// A set of IDs that should never be generated by the generator.
+    ids: HashSet<String>,
+    /// The counter for the current path ID.
+    path: u64,
```

Comment:
Why paths? Aren't you generating paint servers ids?

Reply:
Right now I'm only generating it for the paths of `textPaths`. Paint server IDs are currently not generated anymore.

This means that gradients with object bounding box will still work if text is not converted to paths (since then the `paint_server_to_user_space_on_use` is never called, so the IDs from the paint servers are preserved). But when converting it to paths, it won't work anymore (it didn't work before either, though).

I tried reimplemeting it for paint servers using the same approach as for text paths, but the problem is I can't use the same approach that easily because the order in which we process the paint servers might not be the same. So for example, if I use `loop_over_paint_servers` to find all paint servers without an ID it will loop over paths/texts as well as corresponding underlines/overlines in a specific order, and if I push the new IDs into a vector just like for text paths, I need to make sure that when writing the paint servers I use the exact same order which requires some more thinking on how to do it, so I ignored it for now.

Reply:
>Right now I'm only generating it for the paths of textPaths. 

Aren't they always have an ID?

>the paint servers I use the exact same order

Why do you care about the order in the first place? All you have to do is to have `HashMap<Node, String>`, no?

Reply:
> Aren't they always have an ID?

No, because text paths only store the data of the path they belong to, not their actual ID:

https://github.com/RazrFalcon/resvg/blob/9101f30a70398cf7acd6def11f0743d3fc69757e/crates/usvg-tree/src/text.rs#L240-L248

> Why do you care about the order in the first place? All you have to do is to have HashMap<Node, String>, no?

Well, for example the underline and overline of a text might have a different fill, right? But they belong to the same TextNode (not to mention all other tspans of that Text node who might have their own fills, so that wouldn't work I think?

Reply:
>No, because text paths only store the data of the path they belong to, not their actual ID:

Meh... I guess this is because text layers were never preserved originally, so it was fine. And even if we preserve IDs, we still have to make sure to deduplicate those paths during writing. Because multiple text elements can reference the same path.

>Well, for example the underline and overline of a text might have a different fill, right?

Damn...